rmf_gazebo_plugins/src/readonly.cpp

```diff
@@ -0,0 +1,466 @@
+          level.name.c_str(), level.nav_graphs.size());
+
+      if (_nav_graph_index < level.nav_graphs.size())
+      {
+        _found_graph = true;
+        RCLCPP_INFO(logger(), "Graph index [%d] containts [%d] waypoints",
+            _nav_graph_index,
+            level.nav_graphs[_nav_graph_index].vertices.size());
+        initialize_graph();
+      }
+      else
+      {
+        RCLCPP_ERROR(logger(),
+            "Specified nav_graph index [%d] does not exist in level [%s]",
+            _nav_graph_index, _level_name.c_str());
+      }
+      break;
+    }
+  }
+
+  if (!_found_level)
+    RCLCPP_ERROR(logger(),
+        "Did not find level [%s] in building map. Path will not be published.",
+        _level_name.c_str());
+}
+
+void ReadonlyPlugin::initialize_graph()
+{
+  if (!_found_graph)
+    return;
+  
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  _initialized_graph = false;
+
+  const auto start_time = std::chrono::steady_clock::now();
+  _graph = _level.nav_graphs[_nav_graph_index];
+  RCLCPP_INFO(logger(), "Nav graph contains [%d] lanes" , _graph.edges.size());
+  for (const auto& edge : _graph.edges)
+  {
+    // Inserting entry for v1_idx
+    auto entry = _neighbor_map.find(edge.v1_idx);
+    if (entry == _neighbor_map.end())
+    {
+      // We assume the vertex is a neighbor of itself to account for when the
+      // robot turns around and heads back to the start index
+      std::unordered_set<std::size_t> neighbors({edge.v1_idx, edge.v2_idx});
+      _neighbor_map.insert(std::make_pair(edge.v1_idx, neighbors));
+    }
+    // Updating existing entry
+    else
+    {
+      entry->second.insert(edge.v2_idx);
+    }
+
+    // If bidrectional, add entry for v2_idx
+    if (edge.edge_type == edge.EDGE_TYPE_BIDIRECTIONAL)
+    {
+      entry = _neighbor_map.find(edge.v2_idx);
+      if (entry == _neighbor_map.end())
+      {
+        std::unordered_set<std::size_t> neighbors({edge.v1_idx, edge.v2_idx});
+        _neighbor_map.insert(std::make_pair(edge.v2_idx, neighbors));
+      }
+      // Updating existing entry
+      else
+      {
+        entry->second.insert(edge.v1_idx);
+      }
+    }
+
+  }
+
+  // Initialize _wp_names
+  for (std::size_t i =0; i < _graph.vertices.size(); ++i)
+    _wp_names.insert(std::make_pair(i, _graph.vertices[i].name));
+
+  _initialized_graph = true;
+
+}
+
+double ReadonlyPlugin::compute_ds(
+    const ignition::math::Pose3d& pose,
+    const std::size_t& wp)
+{
+  // TODO consider returning a nullptr instead
+  assert(_found_graph);
+  assert(wp < _graph.vertices.size());
+  
+  ignition::math::Vector3d world_position{pose.Pos().X(), pose.Pos().Y(), 0};
+  ignition::math::Vector3d graph_position{
+    _graph.vertices[wp].x,
+    _graph.vertices[wp].y,
+    0};
+
+  return std::abs((world_position - graph_position).Length());
+}
+
+void ReadonlyPlugin::initialize_start(const ignition::math::Pose3d& pose)
+{
+  if (_initialized_start)
+    return;
+
+  if (!_initialized_graph)
```

Comment:
not critical, but it might be safer to fail the plugin if no graph has been received within a timeout, which will be able to reduce the number of similar checks throughout the plugin. Could possibly use something like `spin_some()` within a timeout inside `Load(..)`

Reply:
That is an interesting point. Right now its setup the way it is to overcome the initial startup delay of loading the world + receiving the building map file. Will continue to think about this for the next update.